package KemyBuffers;

message WhiskerTree {
  optional MemoryRange domain = 1;

  repeated WhiskerTree children = 2;

  optional Whisker leaf = 3;

  optional ConfigRange config = 4;

  optional OptimizationSettings optimizer = 5;
}


message MemoryRange {
    message Array {
        repeated double element = 10;
    }
  optional Memory lower = 11;
  optional Memory upper = 12;
  optional uint32 count = 13;
  repeated Array arrs = 14;
}

message Memory {
  optional double ewma_qlen = 21;
  optional double ewma_arrival_rate = 22;
  optional double ewma_depart_rate = 23;
}

message Whisker {
  optional uint32 generation = 31;
  optional bool drop_f = 32;
  optional MemoryRange domain = 33;
}

message OptimizationSetting {
  optional double min_value = 41;
  optional double max_value = 42;

  optional double min_change = 43;
  optional double max_change = 44;

  optional double multiplier = 45;

  optional double default_value = 46;
}

message OptimizationSettings {
  optional OptimizationSetting drop_f = 51;
}

message Range {
  optional double low = 61;
  optional double high = 62;
}

message ConfigRange {
  optional Range num_senders = 71;
  optional Range bottle_bw = 72;
  optional Range bottle_single_delay = 73;

}

message NetConfig {
  optional uint32 num_senders = 1;
  optional double bottle_bw = 2;
  optional double bottle_single_delay = 3;
}
